Pass SelectionPredicate instead of Filter to storage layer

hongchaodeng opened the pull request:
Depends on #31189 (first commit).

ref: #29888

What?
This PR removes the filtering logic and passes SelectionPredicate to storage layer.

Why?
Filter doesn't provide enough information of and isn't the right abstraction for indexing. We need to pass in SelectionPredicate instead.

``` release-note
etcd3: pass SelectionPredicate instead of Filter to storage layer
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31190)

<!-- Reviewable:end -->

hongchaodeng commented:
/cc @lavalamp

hongchaodeng committed:
pass SelectionPredicate instead of Filter to storage layer

hongchaodeng commented:
All tests passed

hongchaodeng commented:
@lavalamp 
Can you help review this PR? The first commit is from #31189 and Clayton has given "lgtm". The second commit is the change of passing SelectionPredicate.
We would like to iterate on review feedback early.

hongchaodeng commented:
"jenkins verif" failure non-relevant:
```
fatal: unable to access 'https://github.com/mesos/mesos-go/': Failed to connect to github.com port 443: Connection timed out
godep: error downloading dep (github.com/mesos/mesos-go/auth):
```

hongchaodeng commented:
@k8s-bot e2e test this issue: #IGNORE

hongchaodeng commented:
@lavalamp 
Can you help review this PR?
The first commit is from #31189 which has "lgtm" already.
The second commit is the change of passing SelectionPredicate.

hongchaodeng commented:
@lavalamp 
Can you review this PR?

calebamiles commented:
@pwittrock: as this continues the work started in https://github.com/kubernetes/kubernetes/pull/31189, do you think it's reasonable to add a 1.5 milestone label on this PR so we
- don't lose track of it when 1.5 coding opens 19 September
- signal that this PR should only be reviewed once the cycles occupied by 1.4 are free

hongchaodeng commented:
@lavalamp 
This is taking way too long. What's blocking this?

lavalamp commented:
sorry, this wasn't actually assigned to me, so I did not look at it when I was doing reviews.

lavalamp commented on federation/registry/cluster/strategy_test.go:
```diff
@@ -145,7 +146,8 @@ func TestMatchCluster(t *testing.T) {
 		for _, field := range fieldSet {
 			m := MatchCluster(labels.Everything(), field.AsSelector())
-			_, matchesSingle := m.MatchesSingle()
+			p := (*apistorage.SelectionPredicate)(m)
```
Fix the return value of MatchCluster. Do not cast.

lavalamp commented on pkg/registry/generic/matcher.go:
```diff
@@ -48,78 +43,4 @@ func MergeFieldsSets(source fields.Set, fragment fields.Set) fields.Set {
-	// Assert implementations match the interface.
-	_ = Matcher(&SelectionPredicate{})
-)
+type SelectionPredicate storage.SelectionPredicate
```
Why do you have this here? This isn't a real type alias. Change all the references instead of leaving this here.

lavalamp replied:
If you're going to clean that up in the future, then at least leave a todo.

hongchaodeng replied:
This alias could be removed.
I was keeping it here for review's sake -- otherwise there will be a lot more changes.
Will change it shortly

lavalamp commented on pkg/storage/interfaces.go:
```diff
@@ -62,34 +64,13 @@ type MatchValue struct {
-	return true
-}
+// FilterFunc takes an API object and returns true if the object satisfies some requirements.
+type FilterFunc func(obj runtime.Object) bool
```
personally, I'd prefer to leave this as an interface.

lavalamp replied:
(just remove the trigger method if you don't need it any more)

lavalamp replied:
Why are you keeping this type at all instead of making everything use SelectionPredicates?

hongchaodeng replied:
Right. We will use SelectionPredicate everywhere.
The filter is kept here for intermediate stage of change.

hongchaodeng commented:
@lavalamp 
NP.
Last time we discussed, you said you would take supervision over this. It would be great to let you know changes through the interface as well.

lavalamp commented on pkg/storage/interfaces.go:
```diff
@@ -62,34 +64,13 @@ type MatchValue struct {
-type Filter interface {
-	// Filter is a predicate which takes an API object and returns true
-	// if and only if the object should remain in the set.
-	Filter(obj runtime.Object) bool
```
You could rename this method to Matches and then the predicate will implement this interface.

lavalamp commented on pkg/registry/core/node/strategy_test.go:
```diff
@@ -39,7 +40,8 @@ func TestMatchNode(t *testing.T) {
 		for _, field := range fieldSet {
 			m := MatchNode(labels.Everything(), field.AsSelector())
-			_, matchesSingle := m.MatchesSingle()
+			p := (*apistorage.SelectionPredicate)(m)
```
Same comment everywhere

lavalamp commented:
Add todos, then.

On Thu, Sep 22, 2016 at 12:26 PM, Hongchao Deng <notifications@github.com>
wrote:

> *@hongchaodeng* commented on this pull request.
> ------------------------------
>
> In pkg/registry/generic/matcher.go
> <https://github.com/kubernetes/kubernetes/pull/31190>:
>
> > -	// object and true here. This will greatly increase efficiency. You
> -	// must still implement Matches(). Note that key does NOT need to
> -	// include the object's namespace.
> -	MatchesSingle() (key string, matchesSingleObject bool)
> -
> -	// For any known index, if a matcher can match only (a subset) of objects
> -	// that return <value> for a given index, a pair (<index name>, <value>)
> -	// will be returned.
> -	MatcherIndex() []storage.MatchValue
> -}
> -
> -var (
> -	// Assert implementations match the interface.
> -	_ = Matcher(&SelectionPredicate{})
> -)
> +type SelectionPredicate storage.SelectionPredicate
>
> This alias could be removed.
> I was keeping it here for review's sake -- otherwise there will be a lot
> more changes.
> Will change it shortly
> ------------------------------
>
> In pkg/storage/interfaces.go
> <https://github.com/kubernetes/kubernetes/pull/31190>:
>
> > -	// false anyway.
> -	Trigger() []MatchValue
> -}
> -
> -// Everything is a Filter which accepts all objects.
> -var Everything Filter = everything{}
> -
> -// everything is implementation of Everything.
> -type everything struct {
> -}
> -
> -func (e everything) Filter(runtime.Object) bool {
> -	return true
> -}
> +// FilterFunc takes an API object and returns true if the object satisfies some requirements.
> +type FilterFunc func(obj runtime.Object) bool
>
> Right. We will use SelectionPredicate everywhere.
> The filter is kept here for intermediate stage of change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31190>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglorJxmy9LR_Um8k7ln3sAL096HBQks5qstZlgaJpZM4Jqa5z>
> .
>

hongchaodeng commented:
@lavalamp 
Thanks for your feedback.
Addressed the comments:
- remove generic.SelectionPredicate
- Add todo for intermediate FilterFunc

PTAL. Will squash at the end.

hongchaodeng commented:
@k8s-bot kubemark test this issue: #IGNORE

hongchaodeng commented:
@k8s-bot kubemark test this issue: #IGNORE

hongchaodeng commented:
@k8s-bot kubemark test this issue: #33388

The kubemark issue is insane.

hongchaodeng commented:
@k8s-bot kubemark gci e2e test this issue: #33388

hongchaodeng commented:
@k8s-bot kubemark gci e2e test this issue: #33388

hongchaodeng commented:
rebased and tests passed.
@lavalamp PTAL. Rebasing quite a few times.

lavalamp commented:
LGTM
